core/src/main/java/com/linecorp/armeria/client/cookie/DefaultCookieJar.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.cookie;
+
+import static java.util.Objects.requireNonNull;
+
+import java.net.URI;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+
+import com.linecorp.armeria.common.Cookie;
+import com.linecorp.armeria.common.CookieBuilder;
+import com.linecorp.armeria.common.Cookies;
+
+import io.netty.util.NetUtil;
+
+/**
+ * A default in-memory {@link CookieJar} implementation.
+ */
+final class DefaultCookieJar implements CookieJar {
+
+    private final Map<Cookie, Long> store;
+    /**
+     * Used to find cookies for a host that matches a domain. For example, if there is a domain example.com,
+     * host example.com or foo.example.com will match all cookies in that entry.
+     */
+    private final Map<String, Set<Cookie>> filter;
+    private final CookiePolicy cookiePolicy;
+    private final ReentrantLock lock;
+
+    DefaultCookieJar() {
+        this(CookiePolicy.acceptOriginOnly());
+    }
+
+    DefaultCookieJar(CookiePolicy cookiePolicy) {
+        this.cookiePolicy = cookiePolicy;
+        store = new HashMap<>();
+        filter = new HashMap<>();
+        lock = new ReentrantLock();
+    }
+
+    @Override
+    public Cookies get(URI uri) {
+        requireNonNull(uri, "uri");
+        if (store.isEmpty()) {
+            return Cookies.of();
+        }
+        final String host = uri.getHost();
+        final String path = uri.getPath().isEmpty() ? "/" : uri.getPath();
+        final boolean secure = "https".equalsIgnoreCase(uri.getScheme());
```

Comment:
Should we consider the schemes like the following?

- `h2`, `h2c`, `h1` and `h1c` (Can be parsed with `SessionProtocol.find()`)
- `xxx+h2`, etc. (Can be parsed with `Scheme.tryParse()`)

i.e.

```java
final SessionProtocol parsedProtocol;
if (scheme.indexOf('+') >= 0) {
    final Scheme parsedScheme = Scheme.tryParse(scheme);
    parsedProtocol = parsedScheme != null ? parsedScheme.sessionProtocol() : null;
} else {
    parsedProtocol = SessionProtocol.find(scheme);
}
final boolean secure = parsedProtocol != null && parsedProtocol.isTls();
```